### ISSUE TITLE
Unknown layer type: MemoryData (known types: Convolution, Eltwise, LRN, Pooling, Power, ReLU, Sigmoid, Softmax, Split, TanH)

### ISSUE THREAD

**Issue by twmht** [BVLC/caffe#5052]
Please use the [caffe-users list](https://groups.google.com/forum/#!forum/caffe-users) for usage, installation, or modeling questions, or other requests for help.
_Do not post such requests to Issues._ Doing so interferes with the development of Caffe.

Please read the [guidelines for contributing](https://github.com/BVLC/caffe/blob/master/CONTRIBUTING.md) before submitting this issue.

### Issue summary

When reading the prototxt from caffe api, it throws

```
Unknown layer type: MemoryData (known types: Convolution, Eltwise, LRN, Pooling, Power, ReLU, Sigmoid, Softmax, Split, TanH)
```

### Steps to reproduce

link to the static library with windows build

there is a similar question http://stackoverflow.com/questions/30325108/caffe-layer-creation-failure

but I don't know how to make it work with msvc compiler using cmake.

### Your system configuration
Operating system: windows 10
Compiler: msvc 2013
CUDA version (if applicable): 8.0
CUDNN version (if applicable): 5.1

**Comment by willyd** [comment]
@twmht Can you setup a minimal example on github that can reproduce this behavior so that I can clone it and try to reproduce.

**Comment by twmht** [comment]
@willyd 

ok, I will try to do that.

**Comment by twmht** [comment]
@willyd 

I have created a minimal example (https://github.com/twmht/forTesting) to reproduce this problem.

Thanks for your help.

**Comment by twmht** [comment]
@willyd 

Any update?

**Comment by willyd** [comment]
@twmht No update. Sorry I am very busy on other stuff these days. Maybe @lunzueta can help I know he has been successful doing this in the past.

**Comment by lunzueta** [comment]
Hi @twmht. I've forked your code and done some modifications (https://github.com/lunzueta/forTesting). Have a look at the changes. Those changes are required to include Caffe and their dependencies correctly in your project. After applying those changes I can make your program start, load the model and so on (without the problem that you report in this issue). Please, let me know if you can run the program in the same way I can in my PC. 

However, while executing your program I get the following error:
```
W1212 21:15:23.979439 14784 memory_data_layer.cpp:90] MemoryData does not transform array data on Reset()
F1212 21:15:28.485384 14784 memory_data_layer.cpp:110] Check failed: data_ MemoryDataLayer needs to be initialized by calling Reset
```
I guess this error comes from the way you are transforming the cv::Mat data into a format that Caffe can understand, with which I'm not so familiar. Sorry I have limited time at this moment to take a look at this other error in more detail, but I hope at least the previous issue is solved. If so, the other error would correspond to a separate issue, that I guess should be handled in the [caffe-users list](https://groups.google.com/forum/#!forum/caffe-users).

**Comment by willyd** [comment]
@lunzueta Thanks for having a look at this. Your help is always appreciated!

**Comment by twmht** [comment]
@lunzueta 

When I use `find_package('caffe')`, it cat not locate the correct installed path ( something like this `caffe_root/build/install`), instead of pointing to the installed path, it points to `caffe_root` (so it can not find the  in `install/include/caffe/export.hpp`),  so I manually find all the dependencies and link to all of them.

So, How can I get the correct installed path by `find_package('caffe')` ( I have tried the `scripts/build_win.cmd` and use the default installed path) ? Can you show how you installed caffe?

By the way, the code of opencv `Mat` part is from the internet.

Maybe I can test the same code in ubuntu to test if this is the opencv problem.

**Comment by lunzueta** [comment]
Thanks for your kind words @willyd :~)

@twmht I followed the instructions explained here (https://github.com/bvlc/caffe/tree/windows) to install Caffe. Additionally, I have CAFFE_ROOT as an environment variable with the path to the `install` folder of the compiled Caffe, which in my case is `C:\caffe\install`. Let me know if this helps you to find Caffe using `find_package('caffe')`.

Regarding the use of `cv::Mat` input data to work with Caffe models, have a look at [this code](https://github.com/BVLC/caffe/blob/master/examples/cpp_classification/classification.cpp), which does this in a different way to the procedure you are using (and it works).

**Comment by twmht** [comment]
Yup! I had seen the code you provided before, but it did not use MemoryDataLayer.

So I have to specify CAFFE_ROOT to the installed path, then CMAKE can find that? I am not very familiar with CMAKE :(

I will try what you suggest and report my progress on this thread.

**Comment by twmht** [comment]
@lunzueta 

I tried the project you have forked, and tested it on ubuntu 16.04
And I use cmake to build caffe static library, but it still throws `MemoryData` layer not found exception.

However, it can not find all the other layers ( known types are empty, `known types: `) , this is different from the case in windows.

```
I1213 22:49:03.899428 16733 layer_factory.hpp:77] Creating layer gta_frames_input_layer
F1213 22:49:03.899449 16733 layer_factory.hpp:81] Check failed: registry.count(type) == 1 (0 vs. 1) Unknown layer type: MemoryData (known types: )
*** Check failure stack trace: ***
```

**Comment by twmht** [comment]
@lunzueta 

Did you build caffe with shared library?

Since I have tried both static library and shared library, and I got the same exception as yours when link to the shared library #5059

**Comment by lunzueta** [comment]
@twmht 
I built it as static library. By the way, regarding my comment about the `CAFFE_ROOT` environment, I've just checked that it's not necessary to have it in order to make `find_package('caffe')` work. It is related to the inclusion of the following folders in the PATH environment variable:
```
C:\caffe\install\libraries\x64\vc14\bin
C:\caffe\install\libraries\bin
C:\caffe\install\libraries\lib
C:\caffe\install\bin
```

**Comment by VasiliBaranov** [comment]
I had the same issue when trying to compile Caffe (actually, the [PSPNet fork](https://github.com/hszhao/PSPNet)) on Windows with VS 2017. I found the solution here: https://stackoverflow.com/questions/30325108/caffe-layer-creation-failure (more precisely, here: https://stackoverflow.com/a/30394784 ). Even more precisely, the comment "Actually /OPT:NOREF with "Use Library Dependency Inputs" set to true worked for me". 

So one has to set (in Visual Studio) References (Project properties->Configuration Properties->Linker->Optimization->References) to NO (/OPT:NOREF). AND one has to set Use Library Dependency Inputs (Project properties->Configuration Properties->Linker->General) to YES. Just setting /OPT:NOREF didn't work for me. I actually additionally set Link Library Dependencies (Project properties->Configuration Properties->Linker->General) to Yes. Didn't try setting it back, so may be it's also important.

**Comment by TechnikEmpire** [comment]
@VasiliBaranov This is done on the project import the caffe static lib, yes? I hope I'm wrong because it's not working for me, same setup. :(

**Comment by TechnikEmpire** [comment]
Proper solution is to use `/WHOLEARCHIVE:caffe_lib_name.lib`, as all other solutions failed for me and there's some tensorflow people suggesting to use `WHOLEARCHIVE` for static versions of that as well, and frankly it just makes sense.

**Comment by commanderka** [comment]
Thx. /WHOLEARCHIVE:caffe_lib_name.lib made my day
